### PR TITLE
add thresholded rcf

### PR DESCRIPTION
**Pull request by wnbts**
### Description
This pr adds thresholded rcf to replace existing rcf and sketch-based threshold.

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by kaituo** [comment]
Can you create a remote branch and we can open PR to check into code there? Don't want to impact other people's testing. After we are ready, we will check into mainline and and backport to 1.1.

**Comment by wnbts** [comment]
yes, i am using the trcf remote branch.

File: src/main/java/org/opensearch/ad/ml/CheckpointDao.java [hunk 1]
```diff
@@ -483,6 +553,11 @@ private void logFailure(BulkByScrollResponse response, String detectorID) {
                     // avoid possible null pointer exception
                     samples = new ArrayDeque<>();
                 }
+                ThresholdedRandomCutForest trcf = null;
+                if (json.has(ENTITY_TRCF)) {
+                    trcf = toTrcf(json.getAsJsonPrimitive(ENTITY_TRCF).getAsString());
+                }
+                // TODO: convert rcf and threshold to trcf
```
Comment (kaituo) [inline src/main/java/org/opensearch/ad/ml/CheckpointDao.java:560]: Could you put line 560~571 under the else branch? Want to make it explicit that we only expect trcf or rcf/threshold, but not both.

File: src/main/java/org/opensearch/ad/model/Entity.java [hunk 2]
```diff
@@ -278,7 +278,7 @@ private static String normalizedAttributes(SortedMap<String, String> attributes)
     */
     public static Optional<String> getModelId(String detectorId, SortedMap<String, String> attributes) {
         if (attributes.isEmpty()) {
-            return Optional.empty();
+            return Optional.of(detectorId);
```
Comment (kaituo) [inline src/main/java/org/opensearch/ad/model/Entity.java:281]: prefer to use rcf partition 0 as we discussed.  Or you can leave it to me to change.

Comment (wnbts) [reply src/main/java/org/opensearch/ad/model/Entity.java:281]: it's up to you.

File: src/main/java/org/opensearch/ad/ml/CheckpointDao.java (hunk 1 above)
Comment (wnbts) [reply src/main/java/org/opensearch/ad/ml/CheckpointDao.java]: yes, updated it.

**Review by kaituo — APPROVED** [review]

**Pull request merged**